socs/agents/pysmurf_controller/smurf_subprocess_util.py

```diff
@@ -0,0 +1,336 @@
+        S, cfg, **kwargs
+    )
+    biased = np.logical_and.reduce([
+        rfrac_range[0] < bwa.Rfrac,
+        rfrac_range[1] > bwa.Rfrac
+    ])
+    data = {
+        'filepath': bwa.filepath,
+        'biased_total': int(np.sum(biased)),
+        'biased_per_bg': [
+            int(np.sum(biased[bwa.bgmap == bg])) for bg in range(12)
+        ],
+    }
+    arrays = {
+        'Rfrac': bwa.Rfrac, 'responsivity': bwa.Si, 'Rtes': bwa.R0,
+    }
+    quantiles = np.array([15, 25, 50, 75, 85])
+    data['quantiles'] = _process_quantiles(quantiles, arrays)
+    return data
+
+
+def test():
+    """Function for testing subprocess operation"""
+    print("HEREE")
+    return {'test': 10}
+
+
+runnable_funcs = [
+    take_noise, take_iv, run_uxm_setup, run_uxm_relock, take_bias_steps,
+    take_bias_waves, test, take_bgmap
+]
+func_map = {f.__name__: f for f in runnable_funcs}
+
+
+@dataclass
+class RunCfg:
+    func_name: str
+    "name of function to run"
+
+    args: List = field(default_factory=list)
+    "Args to pass to function"
+
+    kwargs: Dict = field(default_factory=dict)
+    "Kwargs to pass to function"
+
+    slot: Optional[int] = None
+    "Slot to setup smurf control for. If None, uses the slot from the environment variable 'SLOT'"
+
+    run_in_main_process: bool = False
+    "If true, run the function in the main process instead of a subprocess."
+
+    def __post_init__(self):
+        if self.func_name not in func_map:
+            raise ValueError(f"Invalid function name: {self.func_name}")
+
+
+@dataclass
+class RunResult:
+    success: bool
+    "True if function returned succesfully"
+
+    return_val: any = None
+    "Return value from function"
+
+    traceback: Optional[str] = None
+    "Traceback if function raised an exception"
+
+
+class FuncProtocol(protocol.ProcessProtocol):
+    """
+    Process protocol for running one of the above functions in a subprocess.
+    After the subprocess is started, the RunCfg object is encoded and sent to
+    the subprocess through stdin. Once the function is finished, the result is
+    encoded and passed back through FD 3.
+    """
+
+    def __init__(self, cfg: RunCfg):
+        self.cfg = cfg
+        self.result = None
+
+    def connectionMade(self):
+        data = encode_dataclass(self.cfg)
+        self.transport.write(data)
+        self.transport.closeStdin()
+
+    def childDataReceived(self, childFD, data):
+        if childFD in [1, 2]:  # stdout or stderr:
+            print(data.decode())
+
+        if childFD == 3:
+            self.result = json.loads(data.decode())
+
+    def processExited(self, status):
+        if self.result is None:
+            raise RuntimeError("No result received from child process")
+
+        self.deferred.callback(self.result)
+
+
+@defer.inlineCallbacks
+def _run_func_in_subprocess_reactor(cfg: RunCfg) -> RunResult:
+    """Helper function for run_func_in_subprocess, that can assume reactor context"""
```

Comment:
Is `run_func_in_subprocess` now `run_smurf_func`? If not, where is this function?